[dtensor][10/n] switch pointwise op to use op strategy

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111278
* __->__ #111234
* #110717
* #109145

As titled, this also handles sth like [Shard(0), Shard(0)] correctly for
pointwise ops, which was previously errored out

wanchaol committed:
[dtensor] switch pointwise op to use op strategy

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/ops/pointwise_ops.py:
```diff
@@ -376,11 +401,116 @@
 
 
+def pointwise_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+    print(f">>>>>>>> op input op schema: {op_schema}")
```
Is this needed?

wanchaol replied:
updated

fduwjj commented on torch/distributed/_tensor/ops/pointwise_ops.py:
```diff
@@ -376,11 +401,116 @@
+                redistribute_cost=redistribute_costs,
+            )
+        )
+    print(f">>>>>>> final pointwise strategy generated: {pointwise_strategy}.")
```
ditto.

fduwjj commented on torch/distributed/_tensor/ops/pointwise_ops.py:
```diff
@@ -376,11 +401,116 @@
 for op in pointwise_ops:
-    register_prop_rule(op, schema_info=RuntimeSchemaInfo(static_kwargkey=["out"]))(
-        pointwise_rule
+    # register_prop_rule(op, schema_info=RuntimeSchemaInfo(static_kwargkey=["out"]))(
+    #     pointwise_rule
+    # )
```
maybe we want to remove them?

wanchaol committed:
Update on "[dtensor][10/n] switch pointwise op to use op strategy"


As titled, this also handles sth like [Shard(0), Shard(0)] correctly for
pointwise ops, which was previously errored out

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/ops/pointwise_ops.py:
```diff
@@ -376,11 +394,109 @@
+                    )
+                )
+                redistribute_costs.append(
+                    generate_redistribute_costs(input_arg, spec_to_follow)
```
Wait, IIUC, you want to use `input_placements` to calculate `redistribute_cost`?

wanchaol replied:
Yeah good catch

fduwjj commented on torch/distributed/_tensor/ops/pointwise_ops.py:
```diff
@@ -376,11 +394,109 @@
+    ), f"no strategy to follow for {op_schema}!"
+    pointwise_strategy = OpStrategy([])
+
+    for placement_strategy in followed_strategy.strategies:
```
Not a strong opinion here, but for better code readability, is it possible that we give the `followed_strategy` a better like, IIUC, this kind of implies how output (op result) can be distributed? For my learning purpose, why do we call it "follow"?

fduwjj replied:
And also, under what circumstances when there is multiple strategies for a followed_strategy?

wanchaol replied:
"follow" means all other args should follow this strategy. There might be cases there're multiple strategies for a graph mode case, for eager there's no case that there's multiple strategies

fduwjj reviewed (approved):
Unblock for now, mainly some questions.

wanchaol committed:
Update on "[dtensor][10/n] switch pointwise op to use op strategy"


As titled, this also handles sth like [Shard(0), Shard(0)] correctly for
pointwise ops, which was previously errored out

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][10/n] switch pointwise op to use op strategy"


As titled, this also handles sth like [Shard(0), Shard(0)] correctly for
pointwise ops, which was previously errored out

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][10/n] switch pointwise op to use op strategy"


As titled, this also handles sth like [Shard(0), Shard(0)] correctly for
pointwise ops, which was previously errored out

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][10/n] switch pointwise op to use op strategy"


As titled, this also handles sth like [Shard(0), Shard(0)] correctly for
pointwise ops, which was previously errored out

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][10/n] switch pointwise op to use op strategy"


As titled, this also handles sth like [Shard(0), Shard(0)] correctly for
pointwise ops, which was previously errored out

[ghstack-poisoned]
